minor: Get meetbot release ready (#46)
balena-io-playground/meetbot: README.md

# Meetbot 

Meetbot is a Google Meet bot that makes meetings frictionless. 

Meetbot aims to eliminate questions in video calls like:

> _1. Can you hear me? Can you hear us?_ Validates audio input of attendees minus the awkward silence.
> _2. Can you resend those links again?_ Meetbot records chat transcript.
> _3. Do you have any notes from the call?_ Also, records voice caption transcript.
> _4. Was the meeting recorded?_ Auto-records meetings. 
> _5. Does anyone have the links shared in the meeting?_ Saves it all to Google Docs. Easy sharing.

Reviewer: ```suggestion
1. _Can you hear me? Can you hear us?_ Validates audio input of attendees minus the awkward silence.
2. _Can you resend those links again?_ Meetbot records chat transcript.
3. _Do you have any notes from the call?_ Also, records voice caption transcript.
4. _Was the meeting recorded?_ Auto-records meetings. 
5. _Does anyone have the links shared in the meeting?_ Saves it all to Google Docs. Easy sharing.
```

And, many more features.

> ![](img/diagram.drawio.png)

Reviewer: diagram still calls it "easter egg"

Author: Yeah, need to fix the diagram @tmigone @20k-ultra Can you tweak the diagram to replace the easter egg with "User audio validation" feature.

_Easter egg as a feature_

Author: Fixed this

## Try it out!

> Running this project is as simple as deploying it to a balenaCloud application. You can do it in just one click by using the button below:

Reviewer: That makes it sound like you wouldn't need a device to deploy to

Author: Copied it from a balenaLabs project, but you are indeed correct.

Reviewer: I guess with most labs projects it's kind of obvious from the start that it focuses running on a device, whereas MeetBot could just as well be another cloud service

## Getting Started

After getting the bot server [up and running](#deployment), it will start listening for requests to provision new meetbots. 

To get a meetbot in your meeting, navigate to the Meetbot dashboard using the [public device URL](https://www.balena.io/docs/learn/manage/actions/#enable-public-device-url) or the port you configured for the server to run on. 

Click the "Join Meeting" button and enter the Google Meet URL in the text box. Click "Join Meeting" and wait for a few seconds for the bot to join. When the bot joins the meeting or prompts to join the meeting, it will post a message to the chat to signal that's it's ready to go.

### Running the bot locally

After cloning the repository, install the dependencies:

```
npm ci
npm start
```

The bot will now be running but functionality is limited until the bot is [authenticated](#authentication). To get a bot to join a Google Meet, head to the Meetbot dashboard available on `http://localhost:8080` and click the `Join` button. Copy the Google Meet URL and paste it into the text field.

## Authentication

> Authentication is needed to:

Reviewer: After you removed instructions, how does the user know how to set up authentication? 
Even with the previous wording I would still be wondering how to get the TOTP token, as that's definitely not straight forward

Author: The authentication section definitely needs more work, @somombo is working to simplify it.

Author: We need to provide a lot of context, and I am considering a new page for authentication

Author: Added more information but still needs work.

1. Record meetings

> 2. Creating Google Docs to save transcripts
> 3. Checking the calendar to join meetings automatically
> 4. Joining the Google Meet automatically without the "Allow User" prompt

Reviewer: ```suggestion
2. Create Google Docs to save transcripts
3. Check the calendar to join meetings automatically
4. Join the Google Meet automatically without the "Allow User" prompt
```

# Meetbot 

> Meetbot is a Google Meet bot that makes your meetings frictionless. Features include eliminating questions in video calls like:

Reviewer 2: I think this section is a bit confusing tbh, or at least it's making it harder to show what features we offer. I like the "eliminating questions" part but I would clearly list the features too.

How about something like this? 

```suggestion
Meetbot is a Google Meet bot that makes your meetings frictionless. We've all been there, we've all heard the same questions being repeated over and over again:

1. _Can you hear me?_
2. _Can you resend those links?_
3. _Do you have the notes from the call?_
4. _Was the meeting recorded?_
5. _Does anyone have the links shared in the meeting?_

Meetbot helps you out by:
1. Validating audio input of attendees minus the awkward silence.
2. Recording chat transcript.
3. Recording voice caption transcript.
4. Auto-recording meetings.
5. Saving it all to Google Docs for easy sharing!
```

Author: Much cleaner, I like it too

> 1. _Can you hear me? Can you hear us?_ Validates audio input of attendees minus the awkward silence.
> 2. _Can you resend those links again?_ Meetbot records chat transcript.
> 3. _Do you have any notes from the call?_ Also, records voice caption transcript.
> 4. _Was the meeting recorded?_ Auto-records meetings.
> 5. _Does anyone have the links shared in the meeting?_ Saves it all to Google Docs. Easy sharing.

Reviewer 2: ```suggestion
```

ts-node src/google/create-token.ts
```

## Development

To add new features to the bot. Fork the repository and add your feature to `src/meetbot/features`. After finishing your changes, [run the bot locally](#running-the-bot-locally) to check if your feature loads and test your changes. 

> To develop the UI, run `npm run dev-ui`. This will spin up the development server. You should then be able to access the dashboard at `http://localhost:3000/`. Note that the data you will see on the tables are actually being mocked by MirageJS, and not actual request is going to the API. This makes it easier to develop both components in an entirely decoupled way.

Reviewer: ```suggestion
To develop the UI, run `npm run dev-ui`. This will spin up the development server. You should then be able to access the dashboard at `http://localhost:3000/`. Note that the data you will see on the tables is only mocked and no actual requests are sent to the API. This makes it easier to develop the UI independently.
```

## Deployment

To activate all the features of meetbot, [authenticate the meetbot](#authentication) and create the `token.json` file before deployment. 

> To deploy on balenaCloud, use the balena CLI and balena push command as stated below. For more information [check out the docs](https://www.balena.io/docs/learn/deploy/deployment/).

Reviewer: Is it just me or is "belenaCloud" make it sound like this would become a cloud deployment?

Author: I focused on productization but kept/left the platform choice biased since it becomes easier for us.

Reviewer: Oh yes, that definitely makes sense. That comment was meant in the same way as the one above. If someone doesn't know about the Balena terminology, then "deploy on balenaCloud" (shouldn't it be "deploy with"?) makes it sound like it would be a cloud hoster.
TBH I don't know a better wording that uses the correct branding and is unambiguous...
Maybe
```suggestion
To deploy to your device with the help of [balenaCloud](https://www.balena.io/cloud/), use [balena CLI](https://github.com/balena-io/balena-cli/blob/master/INSTALL.md)'s `balena push` command as stated below. For more information check out the [balena docs](https://www.balena.io/docs/learn/deploy/deployment/).
```

```
balena push <Name of fleet>
```

> To deploy on a server, create and fill the `.env` file in the root of the project directory using the `.env.example` file and deploy using the Dockerfile present in the root directory of this repository. 

Reviewer: Is this still correct?

Author: Indeed, that's the big issue. I need help over deployment instructions. Maybe we can have a call to improve this. 

Server would still need a .env file since they don't have a balena.yml to help them out. That's the big issue I am fighting. My .env file way was universal but the idea got taken apart in the clinic. I appreciate the feedback and folks were correct too. I don't like creating .env files since they are a big blocker. Atleast balena.yml gives you a nice interface to enter the value on deploy with balena.

Reviewer: I'm working (very low prio) on an extension to the [local UI block](https://github.com/balenablocks/local-ui) to give the user the option to edit env-vars on the device. For that I parse the compose file to get the list of defined env-vars. That's something balena cloud could do, too to give a nicer experience.

As a super simple first step, we could have the UI show a message "you need to set the following env vars..." until the server has validated that all necessary env-vars have been set

## Credits

> Meetbot's google meet voice caption capture is implemented using [dzaman/google-meet](https://github.com/dzaman/google-meet-transcripts)'s stenographer implementation licensed under GNU GPLv3.

Reviewer: really? That means we can't place our code under the Apache license

Author: Another thing to discuss. I saw this and we need to figure it out. Can stenographer be selectively GPLv3?

Reviewer: Lawyers can discuss *ages* about this. But the FSF is pretty clear on that: As soon as you include GPL code, your code must be under a compatible license. Which in this case would only be GPL or AGPL
